File: Web/src/loader.js
class ZaliLoader {
    constructor() {
        this.bus = new ZaliBus();
        this.modules = new Map();
    }

    /**
     * Registers a module.
     * @param {Object} module - Module instance must have a name string and optionally an init() method.
     */
    register(module) {
        if (!module || !module.name) {
            console.error("[zali_loader] Failed to register invalid module:", module);
            return;
        }
        this.modules.set(module.name, module);
        console.log(`[zali_loader] Module registered: ${module.name}`);
    }

    /**
     * Initializes all registered modules.
     */
    init() {
        console.log("[zali_loader] Bootstrapping modules...");

        // Register a system command to retrieve module info
        this.bus.registerCommand('loader', 'get_modules', () => Array.from(this.modules.keys()));

        // Initialize each module
        for (let [name, module] of this.modules) {
            try {
                if (typeof module.init === 'function') {
                    module.init(this);
                    console.log(`[zali_loader] Module initialized successfully: ${name}`);
                }
            } catch (e) {
                console.error(`[zali_loader] Error during initialization of module ${name}:`, e);
            }
        }

        console.log("[zali_loader] Bootstrap finished.");
    }
}
window.ZaliLoader = ZaliLoader;

File: Core/src/loader.rs
use crate::bus::ZaliBus;

pub trait ZaliModule {
    fn name(&self) -> &str;
    fn init(&self, bus: &mut ZaliBus) -> Result<(), String>;
}

pub struct ZaliLoader {
    pub bus: ZaliBus,
    modules: Vec<String>,
}

impl ZaliLoader {
    pub fn new() -> Self {
        Self {
            bus: ZaliBus::new(),
            modules: Vec::new(),
        }
    }

    /// Registers and initializes a module.
    pub fn register_module<M: ZaliModule + 'static>(&mut self, module: M) -> Result<(), String> {
        let name = module.name().to_string();
        module.init(&mut self.bus)?;
        self.modules.push(name);
        Ok(())
    }

    pub fn get_modules(&self) -> Vec<String> {
        self.modules.clone()
    }
}

impl Default for ZaliLoader {
    fn default() -> Self {
        Self::new()
    }
}
